File: java/terasort/1_base/src/main/java/terasort/random/Sort.java
/*
 *  Copyright 2002-2016 Barcelona Supercomputing Center (www.bsc.es)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package terasort.random;

import java.util.LinkedList;

public class Sort {

    private static int NUM_KEYS;
    private static int UNIQUE_KEYS;
    private static int KEY_LENGTH;
    private static int UNIQUE_VALUES;
    private static int VALUE_LENGTH;
    private static int NUM_FRAGMENTS;
    private static int RANDOM_SEED;
    private static int KEYS_PER_FRAGMENT;
    private static int NUM_RANGES;

    private static Fragment[] result;

    public static void main(String args[]) {
        //Get parameters
        if (args.length != 7) {
            System.out.println("[ERROR] Usage: Sort <NUM_KEYS> <UNIQUE_KEYS> <KEY_LENGTH>");
            System.out.println("                    <UNIQUE_VALUES> <VALUE_LENGTH> <NUM_FRAGMENTS> <RANDOM_SEED>");
            System.exit(-1);
        }
        NUM_KEYS = Integer.valueOf(args[0]);
        UNIQUE_KEYS = Integer.valueOf(args[1]);
        KEY_LENGTH = Integer.valueOf(args[2]);
        UNIQUE_VALUES = Integer.valueOf(args[3]);
        VALUE_LENGTH = Integer.valueOf(args[4]);
        NUM_FRAGMENTS = Integer.valueOf(args[5]);
        RANDOM_SEED = Integer.valueOf(args[6]);
        KEYS_PER_FRAGMENT = NUM_KEYS / NUM_FRAGMENTS;
        NUM_RANGES = NUM_FRAGMENTS;

        System.out.println("NUM_KEYS parameter value = " + NUM_KEYS);
        System.out.println("UNIQUE_KEYS parameter value = " + UNIQUE_KEYS);
        System.out.println("KEY_LENGTH parameter value = " + KEY_LENGTH);
        System.out.println("UNIQUE_VALUES parameter value = " + UNIQUE_VALUES);
        System.out.println("VALUE_LENGTH parameter value = " + VALUE_LENGTH);
        System.out.println("NUM_FRAGMENTS parameter value = " + NUM_FRAGMENTS);
        System.out.println("RANDOM_SEED parameter value = " + RANDOM_SEED);
        System.out.println("KEYS_PER_FRAGMENT parameter value = " + KEYS_PER_FRAGMENT);
        System.out.println("NUM_RANGES parameter value = " + NUM_RANGES);

        // Run sort by key app
        long startTime = System.currentTimeMillis();
        Sort sbk = new Sort();
        sbk.run();

        System.out.println("[LOG] Main program finished.");

        // Syncronize result
        // System.out.println("[LOG] Result size = " + result.size());
        // Uncomment the following lines to see the result keys
        for (int i = 0; i < NUM_RANGES; ++i) {
            System.out.println("Range " + i + " result:");
            System.out.println(result[i]);
        }
        long endTime = System.currentTimeMillis();

        System.out.println("[TIMER] Elapsed time: " + (endTime - startTime) + " ms");
    }

    /**
     * Execution method.
     */
    private void run() {
        // Initialize file Names
        System.out.println("[LOG] Initialising filenames for each matrix");
        initializeVariables();

        // Compute result
        System.out.println("[LOG] Computing result");
        Fragment[][] ranges = new Fragment[NUM_FRAGMENTS][NUM_RANGES];
        for (int i = 0; i < NUM_FRAGMENTS; ++i) {
            Fragment fragment = SortImpl.generateFragment(KEYS_PER_FRAGMENT, UNIQUE_KEYS, KEY_LENGTH, UNIQUE_VALUES, VALUE_LENGTH, RANDOM_SEED + (long) 2 * KEYS_PER_FRAGMENT * i);
            int start = 0;
            int step = UNIQUE_KEYS / NUM_RANGES;
            int end = step;
            for (int j = 0; j < NUM_RANGES; ++j) {
                ranges[i][j] = SortImpl.filterTask(fragment, start, end);
                start += step;
                end += step;
            }
        }
        for (int i = 0; i < NUM_RANGES; ++i) {
            // MERGE-REDUCE
            LinkedList<Integer> q = new LinkedList<Integer>();
            for (int j = 0; j < NUM_FRAGMENTS; ++j) {
                q.add(j);
            }
            int x = 0;
            while (!q.isEmpty()) {
                x = q.poll();
                int y;
                if (!q.isEmpty()) {
                    y = q.poll();
                    ranges[x][i] = SortImpl.reduceTask(ranges[x][i], ranges[y][i]);
                    q.add(x);
                }
            }

            result[i] = SortImpl.sortPartition(ranges[x][i]);
        }
        int sortedCount = 0;
        for (int i = 0; i < NUM_RANGES; ++i) {
            sortedCount += result[i].getCount();
        }
        System.out.println("Total Sorted Elements = " + sortedCount);
    }

    /**
     * Variables initialization.
     */
    private void initializeVariables() {
        result = new Fragment[NUM_FRAGMENTS];
    }

}

File: python/pycompss_lib/algorithms/sort/1_base/src/sort.py
#!/usr/bin/python
#
#  Copyright 2002-2018 Barcelona Supercomputing Center (www.bsc.es)
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

# -*- coding: utf-8 -*-

import sys
import time
import pickle
from numpy import *
from collections import defaultdict
from collections import OrderedDict
from pycompss.api.task import task
from pycompss.api.api import compss_wait_on


@task(returns=dict)
def calculate_range(block, fragments, num_range, idx):
    step = num_range / fragments
    idx2 = 0
    partial_result = {}
    for num in block:
        initial = 0
        final = step
        while initial < num_range:
            if num >= initial and num < final:
                partial_result[initial, idx, idx2] = num
            initial += step
            final += step
        idx2 += 1
    return partial_result


@task(returns=set)
def reduce(result):
    return set().union(*result)


def sorting(partial_result, result, fragments, num_range):
    new_dict = minimize_dict(partial_result)
    step = num_range / fragments
    sorted_new_dict = OrderedDict(sorted(iter(new_dict.items()),
                                         key=lambda k_v: k_v[0], reverse=False))
    initial = 0
    final = step
    sizes = []
    while initial < num_range:
        start_pos = 0
        for k, v in sorted_new_dict.items():
            sizes.append(start_pos)
            start_pos += len(v)
        position = 0
        for k, v in sorted_new_dict.items():
            if int(k) >= initial and int(k) < final:
                result.append(sort_in_worker(v, sizes[position]))
                initial += step
                final += step
            position += 1
    for i in range(len(result)):
        result[i] = compss_wait_on(result[i])
    sort = reduce(result)
    return sort


@task(returns=dict)
def sort_in_worker(block, position):
    local_block = sorted(block)
    result = {}
    idx = 0
    for num in local_block:
        curr_pos = position + idx
        result[curr_pos] = num
        idx += 1
    return result


def minimize_dict(partial_result):
    new_dict = defaultdict(list)
    for key, value in list(partial_result.items()):
        value = compss_wait_on(value)
        for val in value:
            new_dict[val[0]].append(value[val])
    return new_dict


def sort(nums_file, fragments, num_range):
    from pycompss.api.api import compss_barrier

    # Read nums from file
    f = open(nums_file, 'r')

    dataset = []
    for line in f:
        dataset.append(asarray([int(x) for x in line.strip().split(" ")]))

    # Flat dataset
    nums = [item for sublist in dataset for item in sublist]

    nums = len(nums)

    if nums / fragments < num_range:
        print('ERROR: num_range should be greater than nums/fragments')
        num_range = int(nums / fragments)
        print("Using num_range: %s" % num_range)

    nums_per_node = nums / fragments
    partial_result = {}
    result = []

    start = time.time()

    keys = [nums[i * nums_per_node:i * nums_per_node + nums_per_node]
            for i in range(fragments)]

    idx = 0
    for k in keys:
        partial_result[idx] = calculate_range(k, fragments, num_range, idx)
        idx += 1

    sorted_nums = sorting(partial_result, result, fragments, num_range)

    compss_barrier()

    print("Elapsed time(s)")
    print(time.time() - start)

    print(sorted_nums)

    '''
    # Save result file
    sorted_nums = compss_wait_on(sorted_nums)
    aux = list(sorted_nums.items())
    result_file = open('./result.txt','w')
    pickle.dump(aux, result_file)
    result_file.close()
    '''


if __name__ == "__main__":
    nums_file = sys.argv[1]
    fragments = int(sys.argv[2])
    num_range = int(sys.argv[3])

    sort(nums_file, fragments, num_range)
